Brokerages/Bitfinex/BitfinexBrokerage.cs

```diff
@@ -0,0 +1,449 @@
+                Price = GetPrice(order).ToString(),
+                Symbol = order.Symbol.Value,
+                Type = MapOrderType(order.Type),
+                Exchange = _exchange,
+                Side = order.Quantity > 0 ? buy : sell
+            };
+
+            var response = _restClient.SendOrder(newOrder);
+
+            if (response != null && response.OrderId != 0)
+            {
+                if (CachedOrderIDs.ContainsKey(order.Id))
+                {
+                    CachedOrderIDs[order.Id].BrokerId.Add(response.OrderId.ToString());
+                }
+                else
+                {
+                    CachedOrderIDs.TryAdd(order.Id, new BitfinexOrder
+                    {
+                        Id = order.Id,
+                        BrokerId = new List<string> { response.OrderId.ToString() },
+                        Price = order.Price / ScaleFactor,
+                        Quantity = totalQuantity * (int)ScaleFactor,
+                        Status = OrderStatus.Submitted,
+                        Symbol = order.Symbol,
+                        Time = order.Time,
+                    });
+                }
+                if (crossOrder != null && crossOrder.Status != OrderStatus.Submitted)
+                {
+                    order.Status = OrderStatus.Submitted;
+                    //switching active order
+                    return PlaceOrder(crossOrder, order);
+                }
+
+                OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, 0, "Bitfinex Order Event") { Status = OrderStatus.Submitted });
+                Log.Trace("BitfinexBrokerage.PlaceOrder(): Order completed successfully orderid:" + order.Id.ToString());
+            }
+            else
+            {
+                //todo: maybe only secondary of cross order failed and order will partially fill.
+                OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, 0, "Bitfinex Order Event") { Status = OrderStatus.Invalid });
+                Log.Trace("BitfinexBrokerage.PlaceOrder(): Order failed Order Id: " + order.Id + " timestamp:" + order.Time + " quantity: " + order.Quantity.ToString());
+                return false;
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Update an existing order
+        /// </summary>
+        /// <param name="order"></param>
+        /// <returns></returns>
+        public override bool UpdateOrder(Orders.Order order)
+        {
+
+            bool hasFaulted = false;
+            foreach (string id in order.BrokerId)
+            {
+                var post = new BitfinexCancelReplacePost
+                {
+                    Amount = (order.Quantity / ScaleFactor).ToString(),
+                    CancelOrderId = int.Parse(id),
+                    Symbol = order.Symbol.Value,
+                    Price = order.Price <= 0 ? order.Id.ToString() : (order.Price * ScaleFactor).ToString(),
```

Comment:
I had not considered the implications of cross zero here. This is a cancel and replace (rather than an update) so the replacing order needs cross zero detection also.